### PR TITLE
Restructuring

### PR DESCRIPTION
**Pull request by PavelBal**

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -5,13 +5,10 @@ export(clrCallStatic)
 export(clrGet)
 export(clrGetConstructors)
 export(clrGetEnumNames)
-export(clrGetExtPtr)
```
Comment (PavelBal) [inline NAMESPACE:8]: Made internal. Not used anywhere, would be obsolete with R6 restrcture.

File: NAMESPACE [hunk 2]
```diff
@@ -5,13 +5,10 @@ export(clrCallStatic)
...
 export(clrGetFields)
-export(clrGetInnerPkgName)
```
Comment (PavelBal) [inline NAMESPACE:10]: This is now `rNETenv$nativePkgName`

File: NAMESPACE [hunk 3]
```diff
@@ -5,13 +5,10 @@ export(clrCallStatic)
...
 export(clrGetLoadedAssemblies)
 export(clrGetMemberSignature)
 export(clrGetMethods)
-export(clrGetNativeLibName)
```
Comment (PavelBal) [inline NAMESPACE:14]: It was the same as `clrGetInnerPkgName`!

File: NAMESPACE [hunk 4]
```diff
@@ -25,20 +22,16 @@ export(clrLoadAssembly)
 export(clrNew)
 export(clrReflect)
 export(clrSet)
-export(clrSetEnumProperty)
```
Comment (PavelBal) [inline NAMESPACE:28]: Was never implemented.

File: NAMESPACE [hunk 5]
```diff
@@ -25,20 +22,16 @@ export(clrLoadAssembly)
...
 export(clrShutdown)
 export(clrToString)
 export(clrTraceback)
-export(clrTypeNameExtPtr)
```
Comment (PavelBal) [inline NAMESPACE:32]: Made internal - do not export.

File: NAMESPACE [hunk 6]
```diff
@@ -25,20 +22,16 @@ export(clrLoadAssembly)
...
 export(clrTypename)
 export(clrVT)
 export(getCurrentConvertedObject)
-export(getNativeLibsPath)
```
Comment (PavelBal) [inline NAMESPACE:36]: Obsolete

File: R/00classDefinitions.R [hunk 7]
```diff
@@ -1,2 +1,5 @@
 # Declares the S4 class that is used to hold references to CLR objects.
-setClass("cobjRef", representation(clrobj = "externalptr", clrtype = "character"), prototype = list(clrobj = NULL, clrtype = "System.Object"))
+setClass("cobjRef",
```
Comment (PavelBal) [inline R/00classDefinitions.R:2]: Just restyle

File: R/net-object.R [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#' #' @title NetObject
```
Comment (PavelBal) [inline R/net-object.R:1]: This is just a dummy file, ignore it.

File: R/rClr-exported.R [hunk 9]
```diff
@@ -1,69 +1,3 @@
-#' Shuts down the current runtime.
-#'
-#' Shuts down the current runtime.
-#'
-#' @return nothing is returned by this function
-#' @export
-clrShutdown <- function() { # TODO: is this even possible given runtime's constraints?
```
Comment (PavelBal) [inline R/rClr-exported.R:7]: Moved to `utilities-runtime`, nothing changed.

File: R/rClr-exported.R [hunk 10]
```diff
@@ -27,43 +17,6 @@ setConvertAdvancedTypes <- function(enable = TRUE) {
...
-#' stopifnot(file.exists(f))
-#' clrLoadAssembly(f)
-#' # Load an assembly from the global assembly cache (GAC)
-#' clrLoadAssembly("System.Windows.Presentation,
-#'   Version=3.5.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")
-#' # The use of partial assembly names is discouraged; nevertheless it is supported
-#' clrLoadAssembly("System.Web.Services")
-#' clrGetLoadedAssemblies()
-#' }
-clrLoadAssembly <- function(name) {
```
Comment (PavelBal) [inline R/rClr-exported.R:52]: Moved to `utilities-runtime`

File: R/rClr-exported.R [hunk 11]
```diff
@@ -27,43 +17,6 @@ setConvertAdvancedTypes <- function(enable = TRUE) {
...
-  result <- .C("rclr_load_assembly", name, PACKAGE = rNETenv$nativePkgName)
-}
-
-#' Gets the inner name used for the package
-#'
-#' Gets the inner name used for the package (rClrMono or rClrMs). This is not intented for use by most users.
-#'
-#' @return the short name of the library currently loaded, depending on the runtime used (Mono or Microsoft .NET)
-#' @export
-clrGetInnerPkgName <- function() {
```
Comment (PavelBal) [inline R/rClr-exported.R:63]: Use `rNETenv$nativePkgName` directly

File: R/rClr-exported.R [hunk 12]
```diff
@@ -98,66 +51,9 @@ clrReflect <- function(clrobj) {
...
-#'
-#' Prints the last CLR exception. This is roughly the equivalent of the traceback function of R.
-#'
-#' @export
-#' @examples
-#' \dontrun{
-#' clrCallStatic("Rclr.TestCases", "ThrowException", 10L) # will be truncated by the Rf_error API
-#' clrTraceback() # prints the full stack trace
-#' }
-clrTraceback <- function() {
```
Comment (PavelBal) [inline R/rClr-exported.R:115]: Moved to `utilities-runtime`

File: R/rClr-exported.R [hunk 13]
```diff
@@ -98,66 +51,9 @@ clrReflect <- function(clrobj) {
...
-#' @param fullname should the full name of the assemblies be returned
-#' @param filenames if TRUE, return a data frame where the second column is the URI (usually file path) of the loaded assembly.
-#' @return the names of loaded CLR assemblies
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' clrGetLoadedAssemblies()
-#' }
-clrGetLoadedAssemblies <- function(fullname = FALSE, filenames = FALSE) {
```
Comment (PavelBal) [inline R/rClr-exported.R:133]: Moved to `utilities-runtime`

File: R/rClr-exported.R [hunk 14]
```diff
@@ -98,66 +51,9 @@ clrReflect <- function(clrobj) {
...
-#' @param assemblyName the name of the assembly
-#' @return the names of the types exported by the assembly
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' clrGetLoadedAssemblies()
-#' clrGetTypesInAssembly("ClrFacade")
-#' }
-clrGetTypesInAssembly <- function(assemblyName) {
```
Comment (PavelBal) [inline R/rClr-exported.R:155]: Moved to `utilites-assembly`

File: R/rClr-exported.R [hunk 15]
```diff
@@ -213,24 +105,7 @@ clrGetProperties <- function(clrobj, contains = "") {
...
-#' @param type CLR Type, or a (character) type name that can be successfully parsed
-#' @return a list of constructor signatures
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' testClassName <- "Rclr.TestObject"
-#' clrGetConstructors(testClassName)
-#' }
-clrGetConstructors <- function(type) {
```
Comment (PavelBal) [inline R/rClr-exported.R:232]: Moved to `utilities-assembly`

File: R/rClr-exported.R [hunk 16]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' clrIs(testObj, "System.Object")
-#' clrIs(testObj, "System.Double")
-#' (testObj <- clrNew("Rclr.TestMethodBinding"))
-#' # Test for interface definitions
-#' clrIs(testObj, "Rclr.ITestMethodBindings")
-#' clrIs(testObj, clrGetType("Rclr.ITestMethodBindings"))
-#' clrIs(testObj, clrGetType("Rclr.TestMethodBinding"))
-#' clrIs(testObj, clrGetType("System.Reflection.Assembly"))
-#' }
-clrIs <- function(obj, type) {
```
Comment (PavelBal) [inline R/rClr-exported.R:323]: Moved to `utilities-net-object`

File: R/rClr-exported.R [hunk 17]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#'   )
-#' }
-#' f(1:3)
-#' f(3)
-#' f("a")
-#' f("a", 3)
-#' f(3, "a")
-#' f(list("a", 3))
-#' }
-clrCall <- function(obj, methodName, ...) {
```
Comment (PavelBal) [inline R/rClr-exported.R:371]: Moved to `utilities-net-object`

File: R/rClr-exported.R [hunk 18]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' testClassName <- "Rclr.TestObject"
-#' testObj <- clrNew(testClassName)
-#' clrReflect(testObj)
-#' clrGet(testObj, "FieldIntegerOne")
-#' clrGet(testClassName, "StaticPropertyIntegerOne")
-#' }
-clrGet <- function(objOrType, name) {
```
Comment (PavelBal) [inline R/rClr-exported.R:393]: Moved to `utilities-net-object`

File: R/rClr-exported.R [hunk 19]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' clrSet(testClassName, "StaticPropertyIntegerOne", 42)
-#'
-#' # Using 'good old' Windows forms to say hello:
-#' clrLoadAssembly("System.Windows.Forms, Version=2.0.0.0,
-#'   Culture=neutral, PublicKeyToken=b77a5c561934e089")
-#' f <- clrNew("System.Windows.Forms.Form")
-#' clrSet(f, "Text", "Hello from '.NET'")
-#' clrCall(f, "Show")
-#' }
-clrSet <- function(objOrType, name, value) {
```
Comment (PavelBal) [inline R/rClr-exported.R:421]: Moved to `utilities-net-object`

File: R/rClr-exported.R [hunk 20]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#'
-#' @param enumType a CLR object, System.Type or type name, possibly namespace and assembly qualified type name, e.g. 'My.Namespace.MyClass,MyAssemblyName'.
-#' @return a character vector of the names for the enum
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' clrGetEnumNames("Rclr.TestEnum")
-#' }
-clrGetEnumNames <- function(enumType) {
```
Comment (PavelBal) [inline R/rClr-exported.R:435]: Moved to `utilities-enum`

File: R/rClr-exported.R [hunk 21]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' Sets the value of an enum field or property of an object or class
-#'
-#' Sets the value of an enum field or property of an object or class
-#'
-#' @param objOrType a CLR object, or type name, possibly namespace and assembly qualified type name, e.g. 'My.Namespace.MyClass,MyAssemblyName'.
-#' @param name the name of a field/property of the object
-#' @param enumtype type name, possibly namespace and assembly qualified type name, e.g. 'My.Namespace.MyClass,MyAssemblyName'.
-#' @param enumval the value to set the field with
-#' @export
-clrSetEnumProperty <- function(objOrType, name, enumtype, enumval) {
```
Comment (PavelBal) [inline R/rClr-exported.R:448]: Removed as this is not implemented in C.

File: R/rClr-exported.R [hunk 22]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-}
-
-#' Gets the external pointer CLR object.
-#'
-#' Gets the external pointer CLR object. End user are unlikely to need this.
-#'
-#' @param clrObject a S4 object of class clrobj
-#' @return the external pointer to the CLR object
-#' @export
-clrGetExtPtr <- function(clrObject) {
```
Comment (PavelBal) [inline R/rClr-exported.R:460]: Made internal.

File: R/rClr-exported.R [hunk 23]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' @return a character string, the type name
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' testClassName <- "Rclr.TestObject"
-#' testObj <- clrNew(testClassName)
-#' clrTypeNameExtPtr(clrGetExtPtr(testObj))
-#' }
-clrTypeNameExtPtr <- function(extPtr) {
```
Comment (PavelBal) [inline R/rClr-exported.R:478]: Made internal.

File: R/rClr-exported.R [hunk 24]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' @return type name
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' testClassName <- "Rclr.TestObject"
-#' testObj <- clrNew(testClassName)
-#' clrTypename(testObj)
-#' }
-clrTypename <- function(clrobj) {
```
Comment (PavelBal) [inline R/rClr-exported.R:498]: Moved to `utilities-net-object`

File: R/rClr-exported.R [hunk 25]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-}
-
-
-#' Gets the name of the native library currently loaded.
-#'
-#' Gets the name of the native library currently loaded. Used only for unit tests.
-#'
-#' @return the name of the native library currently loaded: rClrMs or rClrMono
-#' @export
-clrGetNativeLibName <- function() {
```
Comment (PavelBal) [inline R/rClr-exported.R:510]: Removed - direct access from env.

File: R/rClr-exported.R [hunk 26]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' @param ... additional method arguments passed to .External
-#' @return an object resulting from the call. May be a CLR object, or a native R object for common types. Can be NULL.
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' cTypename <- "Rclr.TestCases"
-#' clrCallStatic(cTypename, "IsTrue", TRUE)
-#' }
-clrCallStatic <- function(typename, methodName, ...) {
```
Comment (PavelBal) [inline R/rClr-exported.R:527]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 27]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#'
-#' @param ... method arguments passed to .External
-#' @return a character message with type information about each argument.
-#' @export
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' peekClrArgs("a", numeric(0))
-#' }
-peekClrArgs <- function(...) {
```
Comment (PavelBal) [inline R/rClr-exported.R:544]: Modev to `utilities-diagnostics`

File: R/rClr-exported.R [hunk 28]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' cTypename <- "Rclr.TestCases"
-#' clrGetStaticMembers(cTypename)
-#' testClassName <- "Rclr.TestObject"
-#' testObj <- clrNew(testClassName)
-#' clrGetStaticMembers(testObj)
-#' }
-clrGetStaticMembers <- function(objOrType) {
```
Comment (PavelBal) [inline R/rClr-exported.R:564]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 29]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-}
-
-#' Gets the static fields for a type
-#'
-#' Gets the static fields for a type
-#'
-#' @param objOrType a CLR object, or type name, possibly namespace and assembly qualified type name, e.g. 'My.Namespace.MyClass,MyAssemblyName'.
-#' @param contains a string that the property names returned must contain
-#' @export
-clrGetStaticFields <- function(objOrType, contains = "") {
```
Comment (PavelBal) [inline R/rClr-exported.R:575]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 30]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-  clrCallStatic(reflectionHelperTypeName, "GetStaticFields", objOrType, contains)
-}
-
-#' Gets the static members for a type
-#'
-#' Gets the static members for a type
-#'
-#' @inheritParams clrGetStaticFields
-#' @export
-clrGetStaticProperties <- function(objOrType, contains = "") {
```
Comment (PavelBal) [inline R/rClr-exported.R:585]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 31]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-  clrCallStatic(reflectionHelperTypeName, "GetStaticProperties", objOrType, contains)
-}
-
-#' Gets the static members for a type
-#'
-#' @inheritParams clrGetStaticFields
-#' @export
-clrGetStaticMethods <- function(objOrType, contains = "") {
```
Comment (PavelBal) [inline R/rClr-exported.R:593]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 32]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-  clrCallStatic(reflectionHelperTypeName, "GetStaticMethods", objOrType, contains)
-}
-
-#' Gets the signature of a static member of a type
-#'
-#' @param typename type name, possibly namespace and assembly qualified type name, e.g. 'My.Namespace.MyClass,MyAssemblyName'.
-#' @param memberName The exact name of the member (i.e. field, property, method) to search for
-#' @export
-clrGetStaticMemberSignature <- function(typename, memberName) {
```
Comment (PavelBal) [inline R/rClr-exported.R:602]: Moved to `utilities-static`

File: R/rClr-exported.R [hunk 33]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' Architecture dependent path to the rClr native library
-#'
-#' Guess the directory where to expect the architecture dependent native library of a specified package
-#' e.g. for the package rClr, ${R_HOME}/library/rClr/libs/x64
-#' This is a utility that is not specific to the rClr package
-#'
-#' @param pkgName the name of a package, e.g. 'rClr'
-#' @return the prospective path in which a native library would be found, e.g. c:/R/library/rClr/libs/x64
-#' @export
-getNativeLibsPath <- function(pkgName) {
```
Comment (PavelBal) [inline R/rClr-exported.R:615]: Removed - not required.

File: R/rClr-exported.R [hunk 34]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-}
-
-#' Get the type code for a SEXP
-#'
-#' Get the type code for a SEXP, as returned by the TYPEOF macro
-#'
-#' @param sexp an R object
-#' @return the type code, an integer, as defined in Rinternals.h
-#' @export
-getSexpType <- function(sexp) {
```
Comment (PavelBal) [inline R/rClr-exported.R:631]: `utilities-diagnostics`

File: R/rClr-exported.R [hunk 35]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' Peek into the structure of R objects 'as seen from C code'
-#'
-#' Inspect one or more R object to get information on its representation in the engine.
-#' This function is mostly useful for R/rClr developers. It is derived from the 'showArgs'
-#' example in the R extension manual
-#'
-#' @param ... one or more R objects
-#' @return NULL. Information is printed, not returned.
-#' @export
-inspectArgs <- function(...) {
```
Comment (PavelBal) [inline R/rClr-exported.R:645]: `utilities-diagnostics`

File: R/rClr-exported.R [hunk 36]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-#' library(rClr)
-#' cTypename <- "Rclr.TestCases"
-#' #         public static bool IsTrue(bool arg)
-#' clrVT(cTypename, "IsTrue", TRUE)
-#' clrVT("System.Convert", "ToInt64", 123L)
-#' clrVT("System.Convert", "ToUInt64", 123L)
-#' }
-#' @return A string
-#' @export
-clrVT <- function(objOrType, methodName, ...) {
```
Comment (PavelBal) [inline R/rClr-exported.R:669]: `utilities-diagnostics`

File: R/rClr-exported.R [hunk 37]
```diff
@@ -297,414 +172,21 @@ getCurrentConvertedObject <- function() {
...
-}
-
-#' Gets the type of a CLR object resulting from converting an R object
-#'
-#' Gets the type of a CLR object resulting from converting an R object. This function is mostly for documentation purposes, but may be of use to end users.
-#'
-#' @param x An R objects
-#' @return A list, with columns including mode, type,class,length and the string of the corresponding CLR type.
-#' @export
-rToClrType <- function(x) {
```
Comment (PavelBal) [inline R/rClr-exported.R:681]: `utilities-diagnostics`

File: R/rClr-internal.R [hunk 38]
```diff
@@ -1,13 +1,3 @@
-# Full type name of the reflection helper the interop code written in C#
-reflectionHelperTypeName <- "Rclr.ReflectionHelper"
```
Comment (PavelBal) [inline R/rClr-internal.R:2]: No global variables any more - everything moved to env.

File: R/rClr-internal.R [hunk 39]
```diff
@@ -1,13 +1,3 @@
...
-
-# Full type name of the main facade to the interop code written in C#
-clrFacadeTypeName <- "Rclr.ClrFacade"
-
-checkIsExtPtr <- function(extPtr) {
```
Comment (PavelBal) [inline R/rClr-internal.R:7]: New function `.validateIsExtPtr`

**Comment by PavelBal** [comment]
I went through all the code and have re-structured it in a similar way our osp packages are organized.

next step should be the actual re-implementation as R6 class, but for that, I would wait for migration to .NEt.

File: R/utilities-assembly.R [hunk 40]
```diff
@@ -0,0 +1,65 @@
...
+#' column is the URI (usually file path) of the loaded assembly. Default is `FALSE`
+#' @return The names of loaded CLR assemblies
+#' @export
+#' @examples
+#' \dontrun{
+#' library(rClr)
+#' clrGetLoadedAssemblies()
+#' }
+clrGetLoadedAssemblies <- function(fullname = FALSE, filenames = FALSE) {
+  assNames <- clrCallStatic(rNETenv$reflectionHelperTypeName, "GetLoadedAssemblyNames", fullname)
```
Comment (Yuri05) [inline R/utilities-assembly.R:41]: `assNames` lol

**Review by Yuri05 — APPROVED** [review]

**Pull request merged**